File: src/bin/day1.rs
use std::collections::HashSet;

fn main() {
  let input = include_str!("../data/day1.txt").lines()
  .map(|l| l.parse::<i32>().unwrap())
  .collect();

  part_1(&input);
  part_2(&input);
}

fn part_1 (input: &Vec<i32>) {
  let freq_a = input.iter().fold(0, |sum, i| sum + i);
  println!("Frequency = {}", freq_a);
}

fn part_2 (input: &Vec<i32>) {
  let mut first_dupe: Option<i32> = None;
  let mut frequencies: HashSet<i32> = HashSet::new();
  let mut freq_b: i32 = 0;
  frequencies.insert(freq_b);
  loop {
    for i in input {
      freq_b += i;
      if first_dupe.is_none() && frequencies.contains(&freq_b) {
        first_dupe = Some(freq_b);
      }
      frequencies.insert(freq_b);
    }

    if first_dupe.is_some() {
      break;
    }
  }
  println!("First duplicate frequency = {:?}", first_dupe);
}

File: day1.py
def part_1(lines):
    fr = 0
    for step in lines:
        fr += int(step)
    print("Final frequency = {}".format(fr))


def part_2(lines):
    dupe = dupe_freq(lines)
    print("Duplicate frequency = {}".format(dupe))


def dupe_freq(lines):
    fr = 0
    freq_seen = {fr}

    while True:
        for step in lines:
            fr += int(step)
            if fr in freq_seen:
                return fr
            freq_seen.add(fr)


def main():
    with open("./day1.txt") as lines:
        lines = list(lines)
        part_1(lines)
        part_2(lines)


if __name__ == '__main__':
    main()
